### PR TITLE
Store the index we've normalisedEOFs to...

### LINKED ISSUE #1306: Optimise normaliseEOFs
`normaliseEOFs` runs from the first cycle to the (current - 1) cycle writing any missing EOFs. This can take a significant amount of time on a queue with many historic roll cycles present. It shouldn't be necessary if we ensure the previous roll cycle has an EOF prior to writing any entries to a new one.

We will still need to keep the corrective action for queues created by old versions that did not ensure this, but we only need to do that scan once.

-- Details to follow --

### PR DESCRIPTION
**Pull request by nicktindall**
... to avoid re-normalising the queue from the beginning each time. Fixes #1306

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java [hunk 1]
```diff
@@ -921,20 +921,20 @@ public void tableStorePut(CharSequence key, long index) {
...
                 longValue = metaStoreMap.get(keyBytes);
                 if (longValue == null) {
-                    longValue = metaStore.acquireValueFor(key, index);
+                    longValue = metaStore.acquireValueFor(key, defaultValue);
                     int length = key.length();
                     HeapBytesStore<byte[]> key2 = HeapBytesStore.wrap(new byte[length]);
                     key2.write(0, keyBytes, 0, length);
                     metaStoreMap.put(key2, longValue);
-                    return null;
+                    return longValue;
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java:937]: I don't know why this was returning null?

**Review by JerryShea — APPROVED** [review]
nice

File: src/test/java/net/openhft/chronicle/queue/impl/single/NormaliseEOFsTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Before
+    @Override
+    public void recordExceptions() {
+        super.recordExceptions();
+        exceptionMap = Jvm.recordExceptions(true);
+        exceptionTracker = ExceptionTracker.create(ExceptionKey::message, ExceptionKey::throwable, Jvm::resetExceptionHandlers, exceptionMap,
+                (key) -> key.level != LogLevel.DEBUG && key.level != LogLevel.PERF,
+                (key) -> key.level() + " " + key.clazz().getSimpleName() + " " + key.message());
+        ignoreException(ex -> true, "Ignore everything");
```
Comment (nicktindall) [inline src/test/java/net/openhft/chronicle/queue/impl/single/NormaliseEOFsTest.java:50]: This is a bit naughty. We probably should add a way to ask for debug exceptions and iterate over the exceptions in order.

**Review by rogersimmons — COMMENTED** [review]
Is this feature needed?
Isn't it enough to walk backwards from the most recent, writing EOFs where missing, then stopping at the first file with an EOF already present?

**Comment by nicktindall** [comment]
> Is this feature needed?
> Isn't it enough to walk backwards from the most recent, writing EOFs where missing, then stopping at the first file with an EOF already present?
> 

Apparently not. We did prototype that and it appeared to work but @alamar [made the point](https://github.com/OpenHFT/Chronicle-Queue/issues/1306#issuecomment-1439896535) that there can be numerous pre-touched (EOF-less) cycles present when earlyAcquireNextCycle is turned on.

I would much prefer to do this without adding state, but until we've made the early acquired cycles distinguishable from actual cycles I fear that's not a possibility. This should be considered a stop gap until we can do that or similar.

**Comment by rogersimmons** [comment]
ok, but isn't this the problem which needs to be fixed then (rather than adding more state to tolerate it)?
"there can be numerous pre-touched (EOF-less) cycles present when earlyAcquireNextCycle is turned on"

**Comment by nicktindall** [comment]
> ok, but isn't this the problem which needs to be fixed then (rather than adding more state to tolerate it)?
> "there can be numerous pre-touched (EOF-less) cycles present when earlyAcquireNextCycle is turned on" 

Yes, agreed

**Review by rogersimmons — CHANGES_REQUESTED** [review]
Could you add an issue (tagged as high priority) to address the root cause and link it back to this ticket pls.
Also wrap all changes on this ticket with an explicit comment so they can easily be removed in due course. (The hash/annotate alone won't be enough as may not survive long).

Will then approve 
Thanks

**Comment by nicktindall** [comment]
> Could you add an issue (tagged as high priority) to address the root cause and link it back to this ticket pls. Also wrap all changes on this ticket with an explicit comment so they can easily be removed in due course. (The hash/annotate alone won't be enough as may not survive long).
> 
> Will then approve Thanks

I created https://github.com/ChronicleEnterprise/Chronicle-Queue-Enterprise/issues/423 which should remove the need for normaliseEOFs altogether. This code will all be deleted naturally then because it's primarily to do with normaliseEOFs.

**Review by rogersimmons — APPROVED** [review]
Thanks Nick. Have approved

**Pull request merged**